Implementation: Frontend afsnit (#136)
SW709E19/Report: report/ChapterImplementation/Frontend.tex

\section{Frontend}\label{cha:implementation-frontend}

> The frontend is responsible for showing information about gateways and devices. This information should correspond to that specified in \autoref{subsec:frontend}.

Reviewer: ```suggestion
The frontend is responsible for showing information about gateways and devices. 
This information should correspond to that specified in \autoref{subsec:frontend}.
```

\subsection{Design} \label{subsec:Frontend-Design}
The design of the frontend is designed with high contrast in mind, allowing for quick identification of important information, such as battery status, configuration button and data visualisation.

> Marked in \textbf{blue} on \autoref{fig:screenshot_frontend}, is the gateways. Only the names of the gateways are shown since the gateways do not contain valuable information. Under the name of the gateway, each device that is connected to that gateway is shown. Devices each have a line in the table.

Reviewer: ```suggestion
Marked in \textbf{blue} on \autoref{fig:screenshot_frontend}, is the gateways. 
Only the names of the gateways are shown since the gateways do not contain valuable information.
Under the name of the gateway, each device that is connected to that gateway is shown. 
Devices each have a line in the table.
```

> Marked in \textbf{green} on \autoref{fig:screenshot_frontend}, is the name of the sensor devices. Marked in \textbf{red} is the current battery status. The battery status is colour coded, to indicate how much time the batteries are expected to last. The different levels are as follows:

Reviewer: ```suggestion
Marked in \textbf{green} on \autoref{fig:screenshot_frontend}, is the name of the sensor devices.
Marked in \textbf{red} is the current battery status.
The battery status is colour coded, to indicate how much time the batteries are expected to last.
The different levels are as follows:
```

\begin{itemize}
 \item > 20 days left will be Green, indicating no need for attention.
 \item > 5 days left will be Yellow, indicating may need attention.
 \item > 3 days left will be Orange, indicating attention.
 \item < 3 days left will be Red, indicating urgent attention.
\end{itemize}

> We made the choice to not use battery percentage, but days left as the indicator because devices could have a low battery percentage, but still last for several weeks, and not needing attention from the maintenance responsible yet. 

Reviewer: ```suggestion
The choice was made to not use battery percentage, but instead use days left as the indicator because devices could have a low battery percentage, but still last for several weeks, and not needing attention from the maintenance responsible yet. 
```

> Marked in \textbf{purple} on \autoref{fig:screenshot_frontend}, is the date and time, the device last sent a measurement to the cloud. Next to this marked in \textbf{white} is the location of the device, telling which building and machine the device is attached to. This is valuable information for the maintenance responsible, since he/she has to locate the devices when a battery change or other maintenance is required. 

Reviewer: ```suggestion
Marked in \textbf{purple} on \autoref{fig:screenshot_frontend}, is the date and time, the device last sent a measurement to the cloud. 
Next to this, marked in \textbf{white}, is the location of the device, telling which building and machine the device is attached to. 
This is valuable information for the maintenance responsible, since he/she has to locate the devices when a battery change or other maintenance is required. 
```

> Marked in \textbf{orange} on \autoref{fig:screenshot_frontend}, is the buttons used to show additional information about the sensor devices. If the blue "configuration" is clicked, the dialogue shown on \autoref{fig:screenshot_configuration} will appear, showing what that sensor device is configured to. It is not possible to make changes to the configuration from the dashboard. In the \textbf{orange} marked area, is also a green "Data" button, clicking this button will show the last 60 seconds of data collected by the device, the graph can be seen on \autoref{fig:screenshot_data}.

Reviewer: ```suggestion
Marked in \textbf{orange} on \autoref{fig:screenshot_frontend}, is the buttons used to show additional information about the sensor devices.
If the blue "configuration" button is clicked, the dialogue shown on \autoref{fig:screenshot_configuration} will appear, showing what the sensor device is configured to.
It is not possible to make changes to the configuration from the dashboard.
In the \textbf{orange} marked area, is also a green "Data" button, clicking this button will show a graph showing the last 60 seconds of data collected by the device, the graph can be seen on \autoref{fig:screenshot_data}.
```

\begin{figure}[H]
 \centering
 \includegraphics[width=1\textwidth]{ChapterImplementation/images/screenshot_colored.png}
 \caption{Screenshot of frontend with colour highlighted information}
 \label{fig:screenshot_frontend}
\end{figure}

\begin{figure}[H]
 \centering
 \includegraphics[width=1\textwidth]{ChapterImplementation/images/screenshot_configuration.png}
 \caption{Screenshot of configuration dialogue, shown when clicking the blue "Configuration" button}
 \label{fig:screenshot_configuration}
\end{figure}

\begin{figure}[H]
 \centering
 \includegraphics[width=1\textwidth]{ChapterImplementation/images/screenshot_data.png}
 \caption{Screenshot of data dialogue, shown when clicking the green "Data" button}
 \label{fig:screenshot_data}
\end{figure}

\subsection{Implementation of frontend}

> The frontend is made as a webpage, rendered from HTML and CSS. The functionality comes from Javascript provided by \textbf{bootstrap} and \textbf{plotly} libraries. The HTML is generated from \textbf{PHP}, which is computed server-side.

Reviewer: ```suggestion
The frontend is made as a webpage, rendered from HTML and CSS.
The functionality comes from Javascript provided by \textbf{bootstrap} and \textbf{plotly} libraries.
The HTML is generated from \textbf{PHP}, which is computed server-side.
```

\subsubsection{Bootstrap}
Bootstrap also called Twitter bootstrap, is an open-source toolkit that is used to create and design webpages.
Bootstrap includes simple functionality such as buttons, dialogues, pop-ups, etc.

> Bootstrap also includes auto-sizing functionality, to make a webpage fit most screen sizes, including phones, laptops, and large TVs.

Reviewer: ```suggestion
Bootstrap also includes auto-sizing functionality, to make a webpage fit most screen sizes, including phones, laptops, and large TVs. \cite{Bootstrap}
```

> Bootstrap was also selected for its ease of use, and because members of the group had previous experience with it. \citep{Bootstrap}

Reviewer: ```suggestion
Bootstrap was also selected for its ease of use, and because members of the group had previous experience with it.
```

\subsubsection{Plotly}

> Plotly is a framework to draw interactable graphs. Plotly is used to show interactable graphs to visualise the data collected by the devices.

Reviewer: ```suggestion
Plotly is a framework to draw interactable graphs. 
Plotly is used to show the interactable graphs visualising the data collected by the devices.
```

The graphs are located in the Data dialogue (see \autoref{subsec:Frontend-Design}), and show the last 60 seconds of data collected from the devices. \citep{Plotly}

\subsubsection{PHP}
PHP is an open-source, object-oriented, server-side programming language, often used to generate HTML for webpages.

> PHP is an interpreted programming language, that is executed upon request.

Reviewer: ```suggestion
PHP is an interpreted programming language, that is executed server side on each request.
```

Because PHP is a server-side language, users can never see what code has been executed, only the result.
PHP is, therefore, a good choice, since the server-side code decide what information is presented to the user, and the user cannot manipulate the code, to gain access to unauthorised information. \citep{PHP}

For PHP to connect to IoT Core, Google provides a PHP library, which makes it easy to pull information about sensor devices and gateways.

> The PHP code is responsible for extracting the information from IoT core and generating the HTML that shows gateways, sensor devices with all its information. The PHP code follows the MVC architecture \citep{MVC}.

Reviewer: ```suggestion
The PHP code is responsible for extracting the information from IoT core and generating the HTML that shows gateways and sensor devices with all its information.
The PHP code follows the Model View Component(MVC) architecture \citep{MVC}.
```

\subsection{Interfaces}

> To allow for a more generic frontend that is not tied to a specific provider such as Google Cloud, there could be built some interfaces, that acts as middle man, between the cloud and the frontend.

Reviewer: ```suggestion
To allow for a more generic frontend that is not tied to a specific provider such as Google Cloud, there could be built some interfaces, that acts as middle layer, between the cloud and the frontend.
```

> This would allow for the same frontend to use different interfaces to access different providers. An architecture where this is implemented can be see on \autoref{fig:screenshot_frontend_backend_interface}.

Reviewer: ```suggestion
This would allow for the same frontend to use different interfaces to access different providers.
An architecture where this is implemented can be see on \autoref{fig:screenshot_frontend_backend_interface}.
```
